File: mycollab-web/src/main/java/com/esofthead/mycollab/module/crm/view/cases/CaseUrlResolver.java
/**
 * This file is part of mycollab-web.
 *
 * mycollab-web is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * mycollab-web is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with mycollab-web.  If not, see <http://www.gnu.org/licenses/>.
 */
package com.esofthead.mycollab.module.crm.view.cases;

import com.esofthead.mycollab.common.UrlEncodeDecoder;
import com.esofthead.mycollab.eventmanager.EventBusFactory;
import com.esofthead.mycollab.module.crm.domain.Account;
import com.esofthead.mycollab.module.crm.events.CaseEvent;
import com.esofthead.mycollab.module.crm.view.CrmUrlResolver;

public class CaseUrlResolver extends CrmUrlResolver {
	public CaseUrlResolver() {
		this.addSubResolver("list", new CaseListUrlResolver());
		this.addSubResolver("add", new CaseAddUrlResolver());
		this.addSubResolver("edit", new CaseEditUrlResolver());
		this.addSubResolver("preview", new CasePreviewUrlResolver());
	}

	public static class CaseListUrlResolver extends CrmUrlResolver {
		@Override
		protected void handlePage(String... params) {
			EventBusFactory.getInstance()
					.post(new CaseEvent.GotoList(this, null));
		}
	}

	public static class CaseAddUrlResolver extends CrmUrlResolver {
		@Override
		protected void handlePage(String... params) {
			EventBusFactory.getInstance().post(
					new CaseEvent.GotoAdd(this, new Account()));
		}
	}

	public static class CaseEditUrlResolver extends CrmUrlResolver {
		@Override
		protected void handlePage(String... params) {
			String decodeUrl = UrlEncodeDecoder.decode(params[0]);
			int caseId = Integer.parseInt(decodeUrl);
			EventBusFactory.getInstance().post(
					new CaseEvent.GotoEdit(this, caseId));
		}
	}

	public static class CasePreviewUrlResolver extends CrmUrlResolver {
		@Override
		protected void handlePage(String... params) {
			String decodeUrl = UrlEncodeDecoder.decode(params[0]);
			int caseId = Integer.parseInt(decodeUrl);
			EventBusFactory.getInstance().post(
					new CaseEvent.GotoRead(this, caseId));
		}
	}
}

File: mycollab-mobile/src/main/java/com/mycollab/mobile/module/crm/view/cases/CaseUrlResolver.kt
package com.mycollab.mobile.module.crm.view.cases

import com.mycollab.common.UrlTokenizer
import com.mycollab.vaadin.EventBusFactory
import com.mycollab.mobile.module.crm.event.CaseEvent
import com.mycollab.mobile.module.crm.event.CrmEvent
import com.mycollab.mobile.module.crm.view.CrmModuleScreenData
import com.mycollab.mobile.module.crm.view.CrmUrlResolver
import com.mycollab.module.crm.domain.CaseWithBLOBs

/**
 * @author MyCollab Ltd
 * @since 6.0.0
 */
class CaseUrlResolver : CrmUrlResolver() {
    init {
        this.addSubResolver("list", CaseListUrlResolver())
        this.addSubResolver("add", CaseAddUrlResolver())
        this.addSubResolver("edit", CaseEditUrlResolver())
        this.addSubResolver("preview", CasePreviewUrlResolver())
    }

    class CaseListUrlResolver : CrmUrlResolver() {
        override fun handlePage(vararg params: String) {
            EventBusFactory.getInstance().post(CrmEvent.GotoActivitiesView(this,
                    CrmModuleScreenData.GotoModule(arrayOf())))
        }
    }

    class CaseAddUrlResolver : CrmUrlResolver() {
        override fun handlePage(vararg params: String) =
                EventBusFactory.getInstance().post(CaseEvent.GotoAdd(this, CaseWithBLOBs()))
    }

    class CaseEditUrlResolver : CrmUrlResolver() {
        override fun handlePage(vararg params: String) {
            val caseId = UrlTokenizer(params[0]).getInt()
            EventBusFactory.getInstance().post(CaseEvent.GotoEdit(this, caseId))
        }
    }

    class CasePreviewUrlResolver : CrmUrlResolver() {
        override fun handlePage(vararg params: String) {
            val caseId = UrlTokenizer(params[0]).getInt()
            EventBusFactory.getInstance().post(CaseEvent.GotoRead(this, caseId))
        }
    }
}